### ISSUE TITLE
How do you deal with word probabilities? does probabilities means frequency in corpus？ 

### ISSUE THREAD

**Issue by Z-YuPeng** [FelixKleineBoesing/FruitFlyEmbeddings#1]
I am also trying to Implement this network in this paper, but I have encountered some problems. I set the vocabulary size to 20000. For a rare word Wi, vi/pi will be a huge number, which will cause the program to output NAN.

**Comment by Z-YuPeng** [comment]
in order to prevent Nan ，i must set lr smaller than 1e-7，but in this time ，the loss won‘t drop down. Why don't the authors of the paper use backpropagation to update the gradient?

**Comment by FelixKleineBoesing** [comment]
Hey, 

It seems to me that you have a bug in the calculation of the word probabilities. Probability means frequency in corpus, that´s correct.
If you use float32, the min number to cover is 1/10e38 which is already pretty huge (in case of decimal places).

**Comment by FelixKleineBoesing** [comment]
I cannot imagine about how they would use backpropagation for this type of network, since the ground truth is part of the input data. You simply have no error that you can propagate back to the input. 
Therefore, they used an energy-based function to capture the dependencies between the context input and target input. I hope that the explanation makes it clear :)

**Comment by Z-YuPeng** [comment]
Thank you! I mean, use the energy function as the loss function to calculate the gradient of w, instead of using the formula in the paper, when I do this, I will not encounter errors，but if i use formula in the paper,i will.
I am wondering if my code is wrong, I am checking my code.
One more point, during the optimization process, do you add gradients or subtract gradients to W? I think for a minimization problem. Should reduce the gradient, but it seems that reducing the gradient can not work.

**Comment by FelixKleineBoesing** [comment]
If you make your repo public, I would have a look at it. Maybe I´ll see something :) 

I am still interested in how you want to apply backpropagate. I´am just curious :)
Regarding the training: I treated the weight updates as gradients and even used adam for optimizing.

**Comment by Z-YuPeng** [comment]
I have uploaded. For code ,I just use pytorch  backward method(),
I also took the partial derivative of the energy function, which seems to be the opposite of the gradient  in the paper.
![IMG_20210317_145526](https://user-images.githubusercontent.com/49511209/111427387-014e7b80-8731-11eb-8923-34322ae4e987.jpg)


https://github.com/ZYPmaster/fruitfly

**Comment by Z-YuPeng** [comment]
I have found the problem because I accumulated the gradient of a batch instead of averaging. Thank you for your help！
